### PR TITLE
[SYCL] Implement parallel_for(range, reduction, func)

### PR DESCRIPTION
**Pull request by v-klochkov**
Currently parallel_for accepting sycl::range may handle only 1 reduction
variable.

Also, this patch had to change/update the methods Reducer::atomic_combine(),
which was the good moment to switch the methods implementation from
the deprecated sycl::atomic class to sycl::ONEAPI::atomic_ref.
The conditions on which the fast-atomics implementations are used were
not changed as that deserves a separate patch.

parallel_for accepting sycl::range works much faster than parallel_for
accepting sycl::nd_range, which means that nd_range version needs some
additional performance tuning soon.

Signed-off-by: Vyacheslav N Klochkov <vyacheslav.n.klochkov@intel.com>

### CODE REVIEW

File: sycl/include/CL/sycl/ONEAPI/reduction.hpp [hunk 1]
```diff
@@ -780,6 +817,217 @@ struct get_reduction_aux_kernel_name_t<sycl::detail::auto_name, Type, B1, B2,
...
+  auto GroupSum = Reduction::getReadWriteLocalAcc(1, CGH);
+  using Name =
+      typename get_reduction_main_kernel_name_t<KernelName, KernelType,
+                                                Reduction::is_usm, false>::name;
+  CGH.parallel_for<Name>(NDRange, [=](nd_item<1> NDId) {
+    // Call user's functions. Reducer.MValue gets initialized there.
+    typename Reduction::reducer_type Reducer;
+    reductionLoop(Range, Reducer, NDId, KernelFunc);
+
+    GroupSum[0] = Reducer.getIdentity();
```
Comment (Pennycook) [inline sycl/include/CL/sycl/ONEAPI/reduction.hpp:849]: I think this should be guarded by `if NdId.get_local_id(0) == 0)`.  Technically speaking this is a race condition, as written.

File: sycl/include/CL/sycl/ONEAPI/reduction.hpp [hunk 2]
```diff
@@ -780,6 +817,217 @@ struct get_reduction_aux_kernel_name_t<sycl::detail::auto_name, Type, B1, B2,
...
+
+  using Name =
+      typename get_reduction_main_kernel_name_t<KernelName, KernelType,
+                                                Reduction::is_usm, false>::name;
+  CGH.parallel_for<Name>(NDRange, [=](nd_item<1> NDId) {
+    // Call user's functions. Reducer.MValue gets initialized there.
+    typename Reduction::reducer_type Reducer;
+    reductionLoop(Range, Reducer, NDId, KernelFunc);
+
+    DoReducePartialSumsInLastWG[0] = 0;
```
Comment (Pennycook) [inline sycl/include/CL/sycl/ONEAPI/reduction.hpp:886]: Same comment as above.

File: sycl/include/CL/sycl/ONEAPI/reduction.hpp [hunk 3]
```diff
@@ -780,6 +817,217 @@ struct get_reduction_aux_kernel_name_t<sycl::detail::auto_name, Type, B1, B2,
...
+  CGH.parallel_for<Name>(NDRange, [=](nd_item<1> NDId) {
+    // Call user's functions. Reducer.MValue gets initialized there.
+    typename Reduction::reducer_type Reducer(Identity, BOp);
+    reductionLoop(Range, Reducer, NDId, KernelFunc);
+
+    // Copy the element to local memory to prepare it for tree-reduction.
+    size_t LID = NDId.get_local_linear_id();
+    LocalReds[LID] = Reducer.MValue;
+    LocalReds[WGSize] = Identity;
+    DoReducePartialSumsInLastWG[0] = 0;
```
Comment (Pennycook) [inline sycl/include/CL/sycl/ONEAPI/reduction.hpp:956]: Same comment as above.

File: sycl/include/CL/sycl/ONEAPI/reduction.hpp [hunk 4]
```diff
@@ -780,6 +817,217 @@ struct get_reduction_aux_kernel_name_t<sycl::detail::auto_name, Type, B1, B2,
...
+        DoReducePartialSumsInLastWG[0] = 1;
+    }
+
+    sycl::detail::workGroupBarrier();
+    if (DoReducePartialSumsInLastWG[0]) {
+      auto LocalSum = Identity;
+      for (size_t I = LID; I < NWorkGroups; I += WGSize)
+        LocalSum = BOp(LocalSum, Reduction::getOutPointer(PartialSums)[I]);
+      LocalReds[LID] = LocalSum;
+      LocalReds[WGSize] = Identity;
```
Comment (Pennycook) [inline sycl/include/CL/sycl/ONEAPI/reduction.hpp:993]: Same comment as above.

File: sycl/include/CL/sycl/handler.hpp [hunk 5]
```diff
@@ -1360,6 +1350,49 @@ class __SYCL_EXPORT handler {
...
+    // to hold local arrays required for the tree-reduction algorithm.
+    constexpr bool IsTreeReduction =
+        !Reduction::has_fast_reduce && !Reduction::has_fast_atomics;
+    size_t OneElemSize =
+        IsTreeReduction ? 0 : sizeof(typename Reduction::result_type);
+#ifdef SYCL_REDUCTION_NUM_EU_THREADS
+    uint32_t NumEUThreads = SYCL_REDUCTION_NUM_EU_THREADS;
+#else
+    uint32_t NumEUThreads = ONEAPI::detail::reduGetMaxNumEUThreads(MQueue);
+#endif
```
Comment (Pennycook) [inline sycl/include/CL/sycl/handler.hpp:1377]: I'd recommend replacing this with a term that isn't tied to Intel GPUs, that's more descriptive of how the variable is used.  Something like "NumConcurrentWorkGroups", maybe?

File: sycl/include/CL/sycl/ONEAPI/reduction.hpp [hunk 6]
```diff
@@ -780,6 +817,217 @@ struct get_reduction_aux_kernel_name_t<sycl::detail::auto_name, Type, B1, B2,
...
+
+template <typename KernelName, typename KernelType, int Dims, class Reduction>
+void reduCGFunc(handler &CGH, KernelType KernelFunc, const range<Dims> &Range,
+                size_t MaxWGSize, uint32_t NumEUThreads, Reduction &Redu) {
+  size_t NWorkItems = Range.size();
+  size_t WGSize = std::min(NWorkItems, MaxWGSize);
+  size_t NWorkGroups = NWorkItems / WGSize;
+  if (NWorkItems % WGSize)
+    NWorkGroups++;
+  size_t MaxNWorkGroups = NumEUThreads;
```
Comment (Pennycook) [inline sycl/include/CL/sycl/ONEAPI/reduction.hpp:1023]: The mapping to Intel GPUs is such that 1 EU thread == 1 sub-group.  It's not clear to me that setting the number of work groups equal to the number of EU threads is particularly meaningful if the work-group size is large.

**Comment by Pennycook** [comment]
@JackAKirk: This PR introduces some usages of `atomic_ref` instead of `atomic`.  Does it address your TODO from line 64?

**Comment by JackAKirk** [comment]
> @JackAKirk: This PR introduces some usages of `atomic_ref` instead of `atomic`. Does it address your TODO from line 64?

No, but I think that is OK since it comes under this:

"The conditions on which the fast-atomics implementations are used were
not changed as that deserves a separate patch."

Basically the 32 bit float case should be covered by IsReduOptForFastAtomicFetch rather than IsReduOptForAtomic64Add.

**Comment by Pennycook** [comment]
> > @JackAKirk: This PR introduces some usages of `atomic_ref` instead of `atomic`. Does it address your TODO from line 64?
> 
> No, but I think that is OK since it comes under this:
> 
> "The conditions on which the fast-atomics implementations are used were
> not changed as that deserves a separate patch."
> 
> Basically the 32 bit float case should be covered by IsReduOptForFastAtomicFetch rather than IsReduOptForAtomic64Add.

Ah, you're right.  I'd missed that in the PR description.  Thanks.

File: sycl/include/CL/sycl/handler.hpp (hunk 5 above)
Comment (v-klochkov) [reply sycl/include/CL/sycl/handler.hpp:1377]: It is unclear if that should be Groups or SubGroups. Would you agree on this?
uint32_t NumConcurrentThreads = ONEAPI::detail::reduGetMaxNumConcurrentThreads();

BTW, I forgot to remove "#ifdef SYCL_REDUCTION_NUM_EU_THREADS" code. It was for debugging, performance runs. 
I am going to either delete that code, OR to add underscores before the macro name and rename it to __SYCL_REDUCTION_NUM_CONCURRENT_THREADS

File: sycl/include/CL/sycl/ONEAPI/reduction.hpp (hunk 1 above)
Comment (v-klochkov) [reply sycl/include/CL/sycl/ONEAPI/reduction.hpp]: Fixed in 808bfa5122a6135f8e6feb45886cc0a59af18528

File: sycl/include/CL/sycl/ONEAPI/reduction.hpp (hunk 2 above)
Comment (v-klochkov) [reply sycl/include/CL/sycl/ONEAPI/reduction.hpp]: Thank you. Fixed in 808bfa5122a6135f8e6feb45886cc0a59af18528

File: sycl/include/CL/sycl/ONEAPI/reduction.hpp (hunk 3 above)
Comment (v-klochkov) [reply sycl/include/CL/sycl/ONEAPI/reduction.hpp]: Fixed in DoReducePartialSumsInLastWG

File: sycl/include/CL/sycl/ONEAPI/reduction.hpp (hunk 4 above)
Comment (v-klochkov) [reply sycl/include/CL/sycl/ONEAPI/reduction.hpp]: Fixed in 808bfa5122a6135f8e6feb45886cc0a59af18528

File: sycl/include/CL/sycl/handler.hpp (hunk 5 above)
Comment (v-klochkov) [reply sycl/include/CL/sycl/handler.hpp]: I used the name suggested by you. 808bfa5122a6135f8e6feb45886cc0a59af18528

File: sycl/include/CL/sycl/ONEAPI/reduction.hpp (hunk 6 above)
Comment (v-klochkov) [reply sycl/include/CL/sycl/ONEAPI/reduction.hpp]: This perhaps can be tuned additionally later. I see now that these heuristics give the best results. Lowering number of work-groups or reducing the size of work-groups gives slower perf.

**Comment by v-klochkov** [comment]
@alexbatashev - please take a quick look at this fix from 'abi breaking' point of view.
I believe I did not add any breaking changes. The reason why I am asking is the test layout_array.cpp, which started giving AST dumps for 'id' class before dumps for 'range'. So, I just changed their order in the test.

**Comment by v-klochkov** [comment]
The corresponding LIT tests are almost ready. I will upload them by the noon/end of Thursday.

**Review by alexbatashev — APPROVED** [review]
ABI changes LGTM

**Review by Pennycook — APPROVED** [review]

File: sycl/include/CL/sycl/ONEAPI/reduction.hpp (hunk 6 above)
Comment (Pennycook) [reply sycl/include/CL/sycl/ONEAPI/reduction.hpp]: Apologies if this comment appears twice -- having some GitHub problems.  I think we should add an explicit TODO in the implementation of MaxNumConcurrentWorkGroups saying that it needs to be tuned for other devices.

Comment (v-klochkov) [reply sycl/include/CL/sycl/ONEAPI/reduction.hpp]: Thank you for the comment. Heuristics definitely need some additional tuning.
There is a comment saying exactly that: https://github.com/v-klochkov/llvm/blob/public_vklochkov_reduction_range_review/sycl/source/detail/reduction.cpp#L57

I'll add "TODO: " to it in a separate [NFC] patch

Comment (Pennycook) [reply sycl/include/CL/sycl/ONEAPI/reduction.hpp]: Yeah, I tried to post this message directly tied to the comment but GitHub wouldn't let me do it!  Adding the TODO in a separate NFC patch sounds good to me.

Comment (v-klochkov) [reply sycl/include/CL/sycl/ONEAPI/reduction.hpp]: I added the TODO comment to this PR: https://github.com/intel/llvm/pull/4361 (commit: 900de465b5b1a281ef65375d6289534e93fb6072)

Comment (v-klochkov) [reply sycl/include/CL/sycl/ONEAPI/reduction.hpp]: I added the TODO comment to this PR: https://github.com/intel/llvm/pull/4361 (commit: 900de465b5b1a281ef65375d6289534e93fb6072)

**Pull request merged**